### ISSUE TITLE
AppImage useability verification

### ISSUE THREAD

**Issue by DevilXD** [DevilXD/TwitchDropsMiner#421]
@guihkx Hey. I've made some changes to the AppImage builder recently, to facilitate building the development version on Python 3.10. The CI seems to pass now, but I'm not entirely sure if the resulting AppImage application will even work.

Since I don't really have a Linux environment to check that, I'd like to kindly ask you to take a look at the changes, and check if the compiled version still works. If you can and in some spare time, of course. Otherwise, well, I guess someone will file a new issue if it turns out not to be working as expected.

Here's a diff: https://github.com/DevilXD/TwitchDropsMiner/compare/f125bf058ea717d0ab57388c81dc4a6f007e44f1...master

**Comment by guihkx** [comment]
Unfortunately, the AppImage broke with those changes:

```
$ ./Twitch.Drops.Miner-x86_64.AppImage
Traceback (most recent call last):
  File "/tmp/.mount_TwitchY8a1jo/usr/src/main.py", line 25, in <module>
    from translate import _
  File "/tmp/.mount_TwitchY8a1jo/usr/src/translate.py", line 7, in <module>
    from utils import json_load, json_save
  File "/tmp/.mount_TwitchY8a1jo/usr/src/utils.py", line 26, in <module>
    import yarl
ModuleNotFoundError: No module named 'yarl'
```

If I had to guess, that error occurs because we're still including Python 3.8 within the AppImage, but the PyPA dependencies were installed with Python 3.10 and are located in a different directory than expected.

You've upgraded the Ubuntu runners from 20.04 (codename `focal`) to 22.04 (codename `jammy`), but forgot to update the repositories and GPG keys on the AppImageBuilder recipe here:

https://github.com/DevilXD/TwitchDropsMiner/blob/8cfa43b52bd82e821107b6370503ddc137dd3e19/appimage/AppImageBuilder.yml#L40-L44

In any case, I wouldn't recommend upgrading to 22.04 anyway (unless absolutely necessary), because the AppImage package (and *I think* the PyInstaller package as well) will stop working for 20.04 users.

And Ubuntu 20.04 will still be officially supported [until 2027](https://wiki.ubuntu.com/Releases), so there will definitely be users of this version for quite a while...

**Comment by DevilXD** [comment]
Okay, so it looks like I fucked it up real bad :sweat: I probably shouldn't be touching this without your guidance, but the development builds have to use Python 3.10 from now on.

I guess I'm reversing the runner back to `ubuntu-20.04`, on both PyInstaller and AppImage builders.

> we're still including Python 3.8 within the AppImage

To resolve issues with SSL certificates (#386), https://github.com/DevilXD/TwitchDropsMiner/commit/91f2dacbcce14d3dbc7519bcbff61d1da5ebcb00 includes `truststore` on both, Windows and Linux versions. For it to work though, Python has to be upgraded from 3.8 to 3.10, which I tried to accomplish with the discussed series of commits.

Truststore seems to be installed, indicating the Python upgrade worked, at least partially:
![picture](https://github.com/DevilXD/TwitchDropsMiner/assets/4180725/7c822dc5-daff-49bb-94ed-835d37079363)

Unfortunately, I am unable to make the `appimage-builder` use `python3.10` instead of whatever else it uses. Nothing in the documentation indicates how the interpreter is chosen, or how it can be changed. Any ideas?

**Comment by guihkx** [comment]
Alright, so if we use Ubuntu 20.04, I think we *must* use `actions/setup-python` to install Python 3.10, or maybe use some [non-official PPA](https://launchpad.net/ubuntu/+ppas).

The problem with the first approach is: We have to figure out a way to also include that version of Python within the final AppImage.

You see, the way AppImageBuilder works, is by including packages from the official Ubuntu repositories. For instance, here:

https://github.com/DevilXD/TwitchDropsMiner/blob/b655803bf5270f0b58ec5fa3d0270e6a2049b841/appimage/AppImageBuilder.yml#L46-L48

AppImageBuilder will look at these two packages ([`gir1.2-appindicator3-0.1`](https://packages.ubuntu.com/focal/gir1.2-appindicator3-0.1), [`python3-tk`](https://packages.ubuntu.com/focal/python3-tk)), look up every single dependency, and include them in the final AppImage.

But because we're still using packages from Ubuntu 20.04 (`focal`) here:

https://github.com/DevilXD/TwitchDropsMiner/blob/b655803bf5270f0b58ec5fa3d0270e6a2049b841/appimage/AppImageBuilder.yml#L40-L44

That means version 3.8 of Python gets included in the final AppImage, and not version 3.10.

Anyway, I'll investigate how `actions/setup-python` works internally, and then I can probably include in the AppImage the Python binaries they're using.

If that doesn't work, then we'll have to stick to plan B (less desirable): We stick to `ubuntu-latest` (Ubuntu 22.04), get rid of `actions/setup-python` altogether (they offer Python 3.10 by default), and then do some small adjustments to the AppImageBuilder script.

The drawback to plan B, like I mentioned earlier, is that the app won't run on Ubuntu 20.04 (and derivative distros), but oh well.

**Comment by DevilXD** [comment]
If it means greater compatibility, i'm good with plan A. [Focal version of `python3-tk`](https://packages.ubuntu.com/focal/python3-tk) indicates that only Python 3.8 is supported though, and I'm assuming that's why only that version gets included. The [Jammy version of `python3-tk`](https://packages.ubuntu.com/jammy/python3-tk) seems to be for Python 3.10, but that's available only on Ubuntu 22.04 =/

Whatever `actions/setup-python` does, it seems to put `python3.10` somewhere on the `PATH`. That *should* be enough to make everything work, assuming the `python3-tk` package somehow derives from the `python3.10` installation, but it doesn't due to the above. Wouldn't including `python3.10` instead solve this? I don't know what happens with tkinter support with that though.

For now, I'm reverting the AppImage section specifically, back to using the default Python 3.8.

**Comment by guihkx** [comment]
Okay, so unfortunately plan B seems to be path of least resistance. 

Internally, `actions/setup-python` downloads a custom Python 3.10 build [from here](https://github.com/actions/python-versions/releases/tag/3.10.13-5997403688) and prepends it to `PATH`.

And while I could technically include that Python build in the AppImage, it's not really an ideal build to ship to users, because all binaries include debug symbols, a bunch of test cases, and other unnecessary files that just makes the whole thing really annoying to clean up before we can include it in the AppImage.

So, since Python 3.10 is pretty much needed to resolve these certificate issues, my proposal is to simply generate the AppImage on Ubuntu 22.04, and then document it that users should have least glibc 2.35 to run the AppImage with no issues.

By the way, this version increase in glibc does not affect the PyInstaller package, which should still work just fine with glibc 2.31.
